### PR TITLE
[FXML-1991] Reciprocal Constant Folding For a Splat Tensor

### PR DESCRIPTION
**Pull request by cmcgirr-amd**
Quantization Operations in TOSA are represnted using the following operators:
```
    %0 = "tosa.const"() {value = dense<3.125000e-02> : tensor<1x1x1x1xf32>} : () -> tensor<1x1x1x1xf32>
    %1 = "tosa.reciprocal"(%0) : (tensor<1x1x1x1xf32>) -> tensor<32x3x224x224xf32>
    %2 = "tosa.mul"(%arg0, %1) {LayerName = "QuantizeLinear_2", shift = 0 : i32} : (tensor<32x3x224x224xf32>, tensor<32x3x224x224xf32>) -> tensor<32x3x224x224xf32>
    %3 = "tosa.cast"(%2) : (tensor<32x3x224x224xf32>) -> tensor<32x3x224x224xi8>
```
Here we apply the reciprocal of a constant scale factor which can be easily folded into the constant.

These changes introduce that folding to help simplify possible pattern matching later on.

### CODE REVIEW

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 1]
```diff
@@ -655,6 +679,9 @@ OpFoldResult SubOp::fold(ArrayRef<Attribute> operands) {
   if (!lhsAttr || !rhsAttr)
     return {};
 
+  if (lhsTy != rhsTy)
```
Comment (mgehre-amd) [inline mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:682]: Does not belong in this PR?

**Review by mgehre-amd — APPROVED** [review]
LGTM!

**Review by TinaAMD — COMMENTED** [review]
This fold isn't upstream yet (in contrast to the sub/add), so I would have suggested to upstream it, but I guess it is also covered by the PR with the general folding for reciprocals, so it might not be worth going through the process for the special case.

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 2]
```diff
@@ -629,6 +629,30 @@ OpFoldResult MulOp::fold(ArrayRef<Attribute> operands) {
...
+    return {};
+  }
+
+  if (!constantAttr.isSplat()) {
+    return {};
+  }
+
+  auto floatVal = constantAttr.getSplatValue<llvm::APFloat>();
+
+  if (!floatVal.isFiniteNonZero()) {
```
[the remark this answers is not in the document]
Comment (cmcgirr-amd) [reply mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:646]: I did not know these are the semantics they used. I will remove the case.

**Comment by cmcgirr-amd** [comment]
> This fold isn't upstream yet (in contrast to the sub/add), so I would have suggested to upstream it, but I guess it is also covered by the PR with the general folding for reciprocals, so it might not be worth going through the process for the special case.

True, https://github.com/Xilinx/llvm-project/pull/29 would solve this. Is this PR getting merged soon? I can also just close this then.

**Review by ljfitz — APPROVED** [review]

**Review by TinaAMD — APPROVED** [review]
LGTM
